File: taiwan_stock_app/backend/app/routers/portfolio.py
"""
Portfolio router
專業級高風險交易分析平台 - 投資組合 API
"""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List, Optional

from app.database import get_db
from app.models import User
from app.schemas.portfolio import (
    PortfolioCreate,
    PortfolioUpdate,
    PortfolioResponse,
    PortfolioListResponse,
    PositionResponse,
    PositionListResponse,
    TransactionCreate,
    TransactionResponse,
    TransactionListResponse,
    PortfolioSummary,
    PositionAllocation,
)
from app.services.portfolio_service import PortfolioService
from app.routers.auth import get_current_user

router = APIRouter(prefix="/api/portfolios", tags=["portfolios"])
portfolio_service = PortfolioService()


# ==================== Portfolio CRUD ====================

@router.post("", response_model=PortfolioResponse)
def create_portfolio(
    request: PortfolioCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """創建投資組合"""
    portfolio = portfolio_service.create_portfolio(
        db=db,
        user_id=current_user.id,
        name=request.name,
        description=request.description,
        initial_capital=request.initial_capital,
    )

    # 計算摘要
    summary = portfolio_service.get_portfolio_summary(db, portfolio.id, current_user.id)

    return PortfolioResponse(
        id=portfolio.id,
        user_id=portfolio.user_id,
        name=portfolio.name,
        description=portfolio.description,
        initial_capital=portfolio.initial_capital,
        created_at=portfolio.created_at,
        updated_at=portfolio.updated_at,
        total_value=summary.get("total_value", 0),
        total_cost=summary.get("total_cost", 0),
        total_pnl=summary.get("total_pnl", 0),
        total_pnl_percent=summary.get("total_pnl_percent", 0),
        positions_count=summary.get("positions_count", 0),
    )


@router.get("", response_model=PortfolioListResponse)
def get_portfolios(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """取得用戶的所有投資組合"""
    portfolios = portfolio_service.get_portfolios(db, current_user.id)

    response_list = []
    for p in portfolios:
        summary = portfolio_service.get_portfolio_summary(db, p.id, current_user.id)
        response_list.append(PortfolioResponse(
            id=p.id,
            user_id=p.user_id,
            name=p.name,
            description=p.description,
            initial_capital=p.initial_capital,
            created_at=p.created_at,
            updated_at=p.updated_at,
            total_value=summary.get("total_value", 0),
            total_cost=summary.get("total_cost", 0),
            total_pnl=summary.get("total_pnl", 0),
            total_pnl_percent=summary.get("total_pnl_percent", 0),
            positions_count=summary.get("positions_count", 0),
        ))

    return PortfolioListResponse(portfolios=response_list, total=len(response_list))


@router.get("/{portfolio_id}", response_model=PortfolioResponse)
def get_portfolio(
    portfolio_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """取得單個投資組合"""
    portfolio = portfolio_service.get_portfolio(db, portfolio_id, current_user.id)
    if not portfolio:
        raise HTTPException(status_code=404, detail="Portfolio not found")

    summary = portfolio_service.get_portfolio_summary(db, portfolio_id, current_user.id)

    return PortfolioResponse(
        id=portfolio.id,
        user_id=portfolio.user_id,
        name=portfolio.name,
        description=portfolio.description,
        initial_capital=portfolio.initial_capital,
        created_at=portfolio.created_at,
        updated_at=portfolio.updated_at,
        total_value=summary.get("total_value", 0),
        total_cost=summary.get("total_cost", 0),
        total_pnl=summary.get("total_pnl", 0),
        total_pnl_percent=summary.get("total_pnl_percent", 0),
        positions_count=summary.get("positions_count", 0),
    )


@router.put("/{portfolio_id}", response_model=PortfolioResponse)
def update_portfolio(
    portfolio_id: int,
    request: PortfolioUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """更新投資組合"""
    portfolio = portfolio_service.update_portfolio(
        db=db,
        portfolio_id=portfolio_id,
        user_id=current_user.id,
        name=request.name,
        description=request.description,
    )
    if not portfolio:
        raise HTTPException(status_code=404, detail="Portfolio not found")

    summary = portfolio_service.get_portfolio_summary(db, portfolio_id, current_user.id)

    return PortfolioResponse(
        id=portfolio.id,
        user_id=portfolio.user_id,
        name=portfolio.name,
        description=portfolio.description,
        initial_capital=portfolio.initial_capital,
        created_at=portfolio.created_at,
        updated_at=portfolio.updated_at,
        total_value=summary.get("total_value", 0),
        total_cost=summary.get("total_cost", 0),
        total_pnl=summary.get("total_pnl", 0),
        total_pnl_percent=summary.get("total_pnl_percent", 0),
        positions_count=summary.get("positions_count", 0),
    )


@router.delete("/{portfolio_id}")
def delete_portfolio(
    portfolio_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """刪除投資組合"""
    success = portfolio_service.delete_portfolio(db, portfolio_id, current_user.id)
    if not success:
        raise HTTPException(status_code=404, detail="Portfolio not found")
    return {"message": "Portfolio deleted"}


# ==================== Transactions ====================

@router.post("/{portfolio_id}/transactions", response_model=TransactionResponse)
def add_transaction(
    portfolio_id: int,
    request: TransactionCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    新增交易記錄

    交易類型：
    - buy: 買入
    - sell: 賣出

    自動更新持倉和計算損益
    """
    transaction = portfolio_service.add_transaction(
        db=db,
        portfolio_id=portfolio_id,
        user_id=current_user.id,
        stock_id=request.stock_id,
        stock_name=request.stock_name,
        transaction_type=request.transaction_type.value,
        quantity=request.quantity,
        price=request.price,
        fee=request.fee,
        tax=request.tax,
        notes=request.notes,
        transaction_date=request.transaction_date,
    )
    if not transaction:
        raise HTTPException(status_code=404, detail="Portfolio not found")

    return TransactionResponse(
        id=transaction.id,
        portfolio_id=transaction.portfolio_id,
        stock_id=transaction.stock_id,
        stock_name=transaction.stock_name,
        transaction_type=transaction.transaction_type.value,
        quantity=transaction.quantity,
        price=transaction.price,
        fee=transaction.fee,
        tax=transaction.tax,
        total_amount=transaction.total_amount,
        notes=transaction.notes,
        transaction_date=transaction.transaction_date,
        created_at=transaction.created_at,
    )


@router.get("/{portfolio_id}/transactions", response_model=TransactionListResponse)
def get_transactions(
    portfolio_id: int,
    limit: int = 50,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """取得交易記錄"""
    transactions = portfolio_service.get_transactions(
        db, portfolio_id, current_user.id, limit
    )

    return TransactionListResponse(
        transactions=[
            TransactionResponse(
                id=t.id,
                portfolio_id=t.portfolio_id,
                stock_id=t.stock_id,
                stock_name=t.stock_name,
                transaction_type=t.transaction_type.value,
                quantity=t.quantity,
                price=t.price,
                fee=t.fee,
                tax=t.tax,
                total_amount=t.total_amount,
                notes=t.notes,
                transaction_date=t.transaction_date,
                created_at=t.created_at,
            )
            for t in transactions
        ],
        total=len(transactions),
    )


# ==================== Positions ====================

@router.get("/{portfolio_id}/positions", response_model=PositionListResponse)
def get_positions(
    portfolio_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """取得持倉列表（自動更新價格）"""
    positions = portfolio_service.update_positions_price(
        db, portfolio_id, current_user.id
    )

    return PositionListResponse(
        positions=[
            PositionResponse(
                id=p.id,
                portfolio_id=p.portfolio_id,
                stock_id=p.stock_id,
                stock_name=p.stock_name,
                quantity=p.quantity,
                avg_cost=p.avg_cost,
                current_price=p.current_price,
                market_value=p.current_price * p.quantity,
                unrealized_pnl=p.unrealized_pnl,
                unrealized_pnl_percent=p.unrealized_pnl_percent,
                last_updated=p.last_updated,
            )
            for p in positions
        ],
        total=len(positions),
    )


# ==================== Statistics ====================

@router.get("/{portfolio_id}/summary", response_model=PortfolioSummary)
def get_portfolio_summary(
    portfolio_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """取得投資組合摘要"""
    summary = portfolio_service.get_portfolio_summary(db, portfolio_id, current_user.id)
    if not summary:
        raise HTTPException(status_code=404, detail="Portfolio not found")

    return PortfolioSummary(**summary)


@router.get("/{portfolio_id}/allocation", response_model=List[PositionAllocation])
def get_position_allocation(
    portfolio_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """取得持倉配置"""
    allocations = portfolio_service.get_position_allocation(
        db, portfolio_id, current_user.id
    )
    return [PositionAllocation(**a) for a in allocations]

File: taiwan_stock_app/frontend/lib/models/portfolio.dart
/// 投資組合數據模型
/// 專業級高風險交易分析平台

/// 投資組合
class Portfolio {
  final int id;
  final int userId;
  final String name;
  final String? description;
  final double initialCapital;
  final DateTime createdAt;
  final DateTime updatedAt;
  final double totalValue;
  final double totalCost;
  final double totalPnl;
  final double totalPnlPercent;
  final int positionsCount;

  Portfolio({
    required this.id,
    required this.userId,
    required this.name,
    this.description,
    required this.initialCapital,
    required this.createdAt,
    required this.updatedAt,
    required this.totalValue,
    required this.totalCost,
    required this.totalPnl,
    required this.totalPnlPercent,
    required this.positionsCount,
  });

  factory Portfolio.fromJson(Map<String, dynamic> json) {
    return Portfolio(
      id: json['id'],
      userId: json['user_id'],
      name: json['name'],
      description: json['description'],
      initialCapital: (json['initial_capital'] as num).toDouble(),
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
      totalValue: (json['total_value'] as num?)?.toDouble() ?? 0,
      totalCost: (json['total_cost'] as num?)?.toDouble() ?? 0,
      totalPnl: (json['total_pnl'] as num?)?.toDouble() ?? 0,
      totalPnlPercent: (json['total_pnl_percent'] as num?)?.toDouble() ?? 0,
      positionsCount: json['positions_count'] ?? 0,
    );
  }

  /// 是否獲利
  bool get isProfitable => totalPnl > 0;

  /// 是否虧損
  bool get isLosing => totalPnl < 0;

  /// 現金餘額
  double get cashBalance => initialCapital - totalCost;
}

/// 投資組合列表響應
class PortfolioListResponse {
  final List<Portfolio> portfolios;
  final int total;

  PortfolioListResponse({
    required this.portfolios,
    required this.total,
  });

  factory PortfolioListResponse.fromJson(Map<String, dynamic> json) {
    return PortfolioListResponse(
      portfolios: (json['portfolios'] as List)
          .map((e) => Portfolio.fromJson(e))
          .toList(),
      total: json['total'],
    );
  }
}

/// 持倉
class Position {
  final int id;
  final int portfolioId;
  final String stockId;
  final String stockName;
  final int quantity;
  final double avgCost;
  final double currentPrice;
  final double marketValue;
  final double unrealizedPnl;
  final double unrealizedPnlPercent;
  final DateTime lastUpdated;

  Position({
    required this.id,
    required this.portfolioId,
    required this.stockId,
    required this.stockName,
    required this.quantity,
    required this.avgCost,
    required this.currentPrice,
    required this.marketValue,
    required this.unrealizedPnl,
    required this.unrealizedPnlPercent,
    required this.lastUpdated,
  });

  factory Position.fromJson(Map<String, dynamic> json) {
    return Position(
      id: json['id'],
      portfolioId: json['portfolio_id'],
      stockId: json['stock_id'],
      stockName: json['stock_name'],
      quantity: json['quantity'],
      avgCost: (json['avg_cost'] as num).toDouble(),
      currentPrice: (json['current_price'] as num).toDouble(),
      marketValue: (json['market_value'] as num).toDouble(),
      unrealizedPnl: (json['unrealized_pnl'] as num).toDouble(),
      unrealizedPnlPercent: (json['unrealized_pnl_percent'] as num).toDouble(),
      lastUpdated: DateTime.parse(json['last_updated']),
    );
  }

  /// 是否獲利
  bool get isProfitable => unrealizedPnl > 0;

  /// 是否虧損
  bool get isLosing => unrealizedPnl < 0;

  /// 成本總額
  double get totalCost => avgCost * quantity;
}

/// 持倉列表響應
class PositionListResponse {
  final List<Position> positions;
  final int total;

  PositionListResponse({
    required this.positions,
    required this.total,
  });

  factory PositionListResponse.fromJson(Map<String, dynamic> json) {
    return PositionListResponse(
      positions: (json['positions'] as List)
          .map((e) => Position.fromJson(e))
          .toList(),
      total: json['total'],
    );
  }
}

/// 交易類型
enum TransactionType {
  buy,
  sell;

  String get displayName {
    switch (this) {
      case TransactionType.buy:
        return '買入';
      case TransactionType.sell:
        return '賣出';
    }
  }

  String get value {
    switch (this) {
      case TransactionType.buy:
        return 'buy';
      case TransactionType.sell:
        return 'sell';
    }
  }

  static TransactionType fromString(String value) {
    switch (value) {
      case 'buy':
        return TransactionType.buy;
      case 'sell':
        return TransactionType.sell;
      default:
        return TransactionType.buy;
    }
  }
}

/// 交易記錄
class Transaction {
  final int id;
  final int portfolioId;
  final String stockId;
  final String stockName;
  final TransactionType transactionType;
  final int quantity;
  final double price;
  final double fee;
  final double tax;
  final double totalAmount;
  final String? notes;
  final DateTime transactionDate;
  final DateTime createdAt;

  Transaction({
    required this.id,
    required this.portfolioId,
    required this.stockId,
    required this.stockName,
    required this.transactionType,
    required this.quantity,
    required this.price,
    required this.fee,
    required this.tax,
    required this.totalAmount,
    this.notes,
    required this.transactionDate,
    required this.createdAt,
  });

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'],
      portfolioId: json['portfolio_id'],
      stockId: json['stock_id'],
      stockName: json['stock_name'],
      transactionType: TransactionType.fromString(json['transaction_type']),
      quantity: json['quantity'],
      price: (json['price'] as num).toDouble(),
      fee: (json['fee'] as num?)?.toDouble() ?? 0,
      tax: (json['tax'] as num?)?.toDouble() ?? 0,
      totalAmount: (json['total_amount'] as num).toDouble(),
      notes: json['notes'],
      transactionDate: DateTime.parse(json['transaction_date']),
      createdAt: DateTime.parse(json['created_at']),
    );
  }

  /// 是否為買入
  bool get isBuy => transactionType == TransactionType.buy;

  /// 是否為賣出
  bool get isSell => transactionType == TransactionType.sell;
}

/// 交易列表響應
class TransactionListResponse {
  final List<Transaction> transactions;
  final int total;

  TransactionListResponse({
    required this.transactions,
    required this.total,
  });

  factory TransactionListResponse.fromJson(Map<String, dynamic> json) {
    return TransactionListResponse(
      transactions: (json['transactions'] as List)
          .map((e) => Transaction.fromJson(e))
          .toList(),
      total: json['total'],
    );
  }
}

/// 投資組合摘要
class PortfolioSummary {
  final double totalValue;
  final double totalCost;
  final double totalPnl;
  final double totalPnlPercent;
  final double cashBalance;
  final int positionsCount;
  final int winningPositions;
  final int losingPositions;
  final String? bestPerformer;
  final String? worstPerformer;

  PortfolioSummary({
    required this.totalValue,
    required this.totalCost,
    required this.totalPnl,
    required this.totalPnlPercent,
    required this.cashBalance,
    required this.positionsCount,
    required this.winningPositions,
    required this.losingPositions,
    this.bestPerformer,
    this.worstPerformer,
  });

  factory PortfolioSummary.fromJson(Map<String, dynamic> json) {
    return PortfolioSummary(
      totalValue: (json['total_value'] as num).toDouble(),
      totalCost: (json['total_cost'] as num).toDouble(),
      totalPnl: (json['total_pnl'] as num).toDouble(),
      totalPnlPercent: (json['total_pnl_percent'] as num).toDouble(),
      cashBalance: (json['cash_balance'] as num).toDouble(),
      positionsCount: json['positions_count'],
      winningPositions: json['winning_positions'] ?? 0,
      losingPositions: json['losing_positions'] ?? 0,
      bestPerformer: json['best_performer'],
      worstPerformer: json['worst_performer'],
    );
  }

  /// 勝率
  double get winRate {
    final total = winningPositions + losingPositions;
    return total > 0 ? winningPositions / total * 100 : 0;
  }
}

/// 持倉配置
class PositionAllocation {
  final String stockId;
  final String stockName;
  final double marketValue;
  final double weight;

  PositionAllocation({
    required this.stockId,
    required this.stockName,
    required this.marketValue,
    required this.weight,
  });

  factory PositionAllocation.fromJson(Map<String, dynamic> json) {
    return PositionAllocation(
      stockId: json['stock_id'],
      stockName: json['stock_name'],
      marketValue: (json['market_value'] as num).toDouble(),
      weight: (json['weight'] as num).toDouble(),
    );
  }
}

/// 創建投資組合請求
class CreatePortfolioRequest {
  final String name;
  final String? description;
  final double initialCapital;

  CreatePortfolioRequest({
    required this.name,
    this.description,
    required this.initialCapital,
  });

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      if (description != null) 'description': description,
      'initial_capital': initialCapital,
    };
  }
}

/// 投資組合詳情（含持股列表）
class PortfolioDetail {
  final int id;
  final String name;
  final String? description;
  final double totalValue;
  final double totalCost;
  final double totalPnl;
  final double totalPnlPercent;
  final List<PortfolioHolding> holdings;

  PortfolioDetail({
    required this.id,
    required this.name,
    this.description,
    required this.totalValue,
    required this.totalCost,
    required this.totalPnl,
    required this.totalPnlPercent,
    required this.holdings,
  });

  factory PortfolioDetail.fromPortfolioAndPositions(
    Portfolio portfolio,
    List<Position> positions,
  ) {
    return PortfolioDetail(
      id: portfolio.id,
      name: portfolio.name,
      description: portfolio.description,
      totalValue: portfolio.totalValue,
      totalCost: portfolio.totalCost,
      totalPnl: portfolio.totalPnl,
      totalPnlPercent: portfolio.totalPnlPercent,
      holdings: positions.map((p) => PortfolioHolding.fromPosition(p)).toList(),
    );
  }
}

/// 投資組合持股
class PortfolioHolding {
  final int id;
  final String stockId;
  final String? stockName;
  final int quantity;
  final double avgCost;
  final double? currentPrice;
  final double? marketValue;
  final double? unrealizedPnl;
  final double? unrealizedPnlPercent;

  PortfolioHolding({
    required this.id,
    required this.stockId,
    this.stockName,
    required this.quantity,
    required this.avgCost,
    this.currentPrice,
    this.marketValue,
    this.unrealizedPnl,
    this.unrealizedPnlPercent,
  });

  factory PortfolioHolding.fromPosition(Position position) {
    return PortfolioHolding(
      id: position.id,
      stockId: position.stockId,
      stockName: position.stockName,
      quantity: position.quantity,
      avgCost: position.avgCost,
      currentPrice: position.currentPrice,
      marketValue: position.marketValue,
      unrealizedPnl: position.unrealizedPnl,
      unrealizedPnlPercent: position.unrealizedPnlPercent,
    );
  }
}

/// 創建交易請求
class CreateTransactionRequest {
  final String stockId;
  final String stockName;
  final TransactionType transactionType;
  final int quantity;
  final double price;
  final double fee;
  final double tax;
  final String? notes;
  final DateTime? transactionDate;

  CreateTransactionRequest({
    required this.stockId,
    required this.stockName,
    required this.transactionType,
    required this.quantity,
    required this.price,
    this.fee = 0,
    this.tax = 0,
    this.notes,
    this.transactionDate,
  });

  Map<String, dynamic> toJson() {
    return {
      'stock_id': stockId,
      'stock_name': stockName,
      'transaction_type': transactionType.value,
      'quantity': quantity,
      'price': price,
      'fee': fee,
      'tax': tax,
      if (notes != null) 'notes': notes,
      if (transactionDate != null)
        'transaction_date': transactionDate!.toIso8601String(),
    };
  }
}
